### PR TITLE
[BYOC] Add example of Composite + Annotate for DNNL fused op

### PR DESCRIPTION
**Pull request by masahi**
This is a reimplementation of #4741 based on the new annotator support for composite added in #5261. This is the first real use case of composite in the code base. It is purely for demonstration purpose and not intended to be used for performance critical scenarios. 

Due to the manually inlined tensors in c codegen, full mobilenet execution is disabled in the test. I tried but it took more than 5GB of RAM and compilation didn't finish in a reasonable time. The related issue discussed in https://discuss.tvm.ai/t/external-codegen-constant-tensors-in-c-codegen/5890/

### CODE REVIEW

**Comment by masahi** [comment]
Please review @zhiics @mbaret @comaniac @trevor-m

File: tests/python/relay/test_pass_partition_graph.py [hunk 1]
```diff
@@ -856,6 +857,111 @@ def expected():
...
+        data = relay.var("data", relay.TensorType((1, 3, 224, 224), "float32"))
+        layer1 = get_blocks("layer1_", data, 3, 8, include_bn, include_sigmoid)
+        layer2 = get_blocks("layer2_", layer1, 8, 8, include_bn, include_sigmoid)
+        return relay.Function(relay.analysis.free_vars(layer2), layer2)
+
+    def get_partitoned_mod(mod, params):
+        # This is required for constant folding
+        mod["main"] = bind_params_by_name(mod["main"], params)
+        pattern_table = [
+            ("dnnl.conv_bias_relu", make_pattern())
```
Comment (masahi) [inline tests/python/relay/test_pass_partition_graph.py:901]: Maybe it is better to move this pattern to `contrib/dnnl.py`. Let me know.

**Comment by masahi** [comment]
@tqchen I got segfault in CI that is not related to this PR. Can you have a look?
https://ci.tvm.ai/blue/organizations/jenkins/tvm/detail/PR-5272/1/pipeline

File: tests/python/relay/test_pass_partition_graph.py [hunk 2]
```diff
@@ -856,6 +857,111 @@ def expected():
...
+        remove_bn_pass = transform.Sequential([
+            transform.InferType(),
+            transform.SimplifyInference(),
+            transform.FoldConstant(),
+            transform.FoldScaleAxis(),
+        ])
+        composite_partition = transform.Sequential([
+            remove_bn_pass,
+            transform.MergeComposite(pattern_table),
+            transform.AnnotateTarget("dnnl"),
```
Comment (alexwong) [inline tests/python/relay/test_pass_partition_graph.py:912]: I'm just curious but is it possible to try MergeCompilerRegions here? In my own testing, I seem to be having issues with composite functions and that pass.

Comment (masahi) [reply tests/python/relay/test_pass_partition_graph.py:912]: I'm not familiar with what MergeCompilerRegions does, but assuming that it merges consecutive ops that are supported by the same compiler, yes it is possible in this case because all ops involved in the composite above, conv2d, add, and relu, are supported by the dnnl backend https://github.com/apache/incubator-tvm/blob/master/python/tvm/relay/op/contrib/dnnl.py#L59-L64. 

But I think composite is supposed to be more general, it works even if `_register_external_op_helper("add")` is missing in dnnl.py above, for example. The purpose of composite is to target a "complex" op whose ops that constitute it are not necessary supported by the compiler by itself.

Comment (mbaret) [reply tests/python/relay/test_pass_partition_graph.py:912]: @alexwong Could you give some more detail on the issue you're facing? I'd be happy to take a look.

Comment (masahi) [reply tests/python/relay/test_pass_partition_graph.py:912]: @alexwong I'll add another pattern in the test, conv2d + relu. I can try MergeCompilerRegions + AnnotateTarget passes in that example. This way we can demonstrate both approaches.

Comment (masahi) [reply tests/python/relay/test_pass_partition_graph.py:912]: @mbaret If I choose MergeCompilerRegions + AnnotateTarget approach, can I still detect fused op inside codegen easily, like composite case?

Comment (mbaret) [reply tests/python/relay/test_pass_partition_graph.py:912]: You can't. Part of the benefit of mergecomposite comes from wrapping the patterns in an easy-to-detect function.

Comment (masahi) [reply tests/python/relay/test_pass_partition_graph.py:912]: Thanks. @alexwong So the answer is both approaches are possible, but composite is more general and it also enables easily detecting fused ops inside codegen, like I do in this line https://github.com/apache/incubator-tvm/pull/5272/files#diff-1defdbd8a4c2ab55fb62ad44e9b314a8R256

Comment (mbaret) [reply tests/python/relay/test_pass_partition_graph.py:912]: The motivation behind MergeCompilerRegions is when you have a codegen that wants to do its own graph-level optimisations, so it should see the full offloaded subgraph rather than one op at a time.

Comment (alexwong) [reply tests/python/relay/test_pass_partition_graph.py:912]: > @alexwong Could you give some more detail on the issue you're facing? I'd be happy to take a look.

An example is I have a very simple Relay graph here where conv2d and clip are supported ops and squeeze and reshape are supported only if it's part of a squeeze->reshape pattern, so I am using mergecomposite for that.

```
def @main(%data: Tensor[(1, 3, 224, 224), float32], %weight1: Tensor[(32, 3, 3, 3), float32]) -> Tensor[(3, 3, 3, 32), float32] {
  %0 = nn.conv2d(%data, %weight1, padding=[1, 1, 1, 1], kernel_size=[3, 3]) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %1 = clip(%0, a_min=0f, a_max=6f) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %2 = squeeze(%1) /* ty=Tensor[(32, 224, 224), float32] */;
  reshape(%2, newshape=[3, 3, 3, 32]) /* ty=Tensor[(3, 3, 3, 32), float32] */
}
```

After annotation I get:
```
def @main(%data: Tensor[(1, 3, 224, 224), float32], %weight1: Tensor[(32, 3, 3, 3), float32]) -> Tensor[(3, 3, 3, 32), float32] {
  %0 = annotation.compiler_begin(%data, meta[relay.attrs.CompilerAttrs][0]) /* ty=Tensor[(1, 3, 224, 224), float32] */;
  %1 = annotation.compiler_begin(%weight1, meta[relay.attrs.CompilerAttrs][1]) /* ty=Tensor[(32, 3, 3, 3), float32] */;
  %2 = nn.conv2d(%0, %1, padding=[1, 1, 1, 1], kernel_size=[3, 3]) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %3 = annotation.compiler_end(%2, meta[relay.attrs.CompilerAttrs][2]) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %4 = annotation.compiler_begin(%3, meta[relay.attrs.CompilerAttrs][3]) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %5 = clip(%4, a_min=0f, a_max=6f) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %6 = annotation.compiler_end(%5, meta[relay.attrs.CompilerAttrs][4]) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %7 = annotation.compiler_begin(%6, meta[relay.attrs.CompilerAttrs][5]) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %9 = fn (%x: Tensor[(1, 32, 224, 224), float32], Composite="compilername.squeeze") -> Tensor[(3, 3, 3, 32), float32] {
    %8 = squeeze(%x) /* ty=Tensor[(32, 224, 224), float32] */;
    reshape(%8, newshape=[3, 3, 3, 32]) /* ty=Tensor[(3, 3, 3, 32), float32] */
  };
  %10 = %9(%7) /* ty=Tensor[(3, 3, 3, 32), float32] */;
  annotation.compiler_end(%10, meta[relay.attrs.CompilerAttrs][6]) /* ty=Tensor[(3, 3, 3, 32), float32] */
}
```

This looks right to me but mergecompilerregions is throwing the following error: 
```
tvm._ffi.base.TVMError: TVMError: Cannot find the corresponding region for start annotation:
v0.0.4
free_var %x: Tensor[(1, 32, 224, 224), float32]
%0 = annotation.compiler_begin(%x, meta[relay.attrs.CompilerAttrs][0]);
%1 = squeeze(%0);
%2 = annotation.compiler_end(%1, meta[relay.attrs.CompilerAttrs][1]);
annotation.compiler_begin(%2, meta[relay.attrs.CompilerAttrs][2])
// meta data omitted. you can use show_meta_data=True to include meta data
```

Not sure if it's expected to not work with composite functions but I think there may be some missing logic there.

Comment (mbaret) [reply tests/python/relay/test_pass_partition_graph.py:912]: Yep, that looks like a bug :) I'll take a quick look now, but I'll probably have a fix PR up tomorrow.

Comment (alexwong) [reply tests/python/relay/test_pass_partition_graph.py:912]: That'd be great! Let me know your initial findings, might be something small :)

File: src/relay/backend/contrib/dnnl/codegen.cc [hunk 3]
```diff
@@ -30,14 +30,108 @@
...
+const CallNode* GetRootConv2DCall(const CallNode* relu_call) {
+  CHECK(relu_call && IsOp(relu_call, "nn.relu"));
+  const auto relu_arg = relu_call->args[0];
+  const CallNode* add_call = relu_arg.as<CallNode>();
+  CHECK(add_call && IsOp(add_call, "add"));
+  const auto add_arg = add_call->args[0];
+  const CallNode* conv_call = add_arg.as<CallNode>();
+  CHECK(conv_call && IsOp(conv_call, "nn.conv2d"));
+  return conv_call;
+}
```
Comment (alexbooth) [inline src/relay/backend/contrib/dnnl/codegen.cc:54]: Maybe this can be more generic, might get tedious to implement this kind of function if there are many different patterns. Could we just pass in a vector of op names like ["nn.relu", "add", "nn.conv2d"] along with the CallNode?

File: src/relay/backend/contrib/dnnl/codegen.cc [hunk 4]
```diff
@@ -133,83 +209,100 @@ class CodegenDNNL : public ExprVisitor, public CodegenCBase {
...
-    auto ishape = GetShape(call->args[0]->checked_type());
-    auto wshape = GetShape(call->args[1]->checked_type());
-
-    // Args: N, C, O
-    args.push_back(std::to_string(ishape[0]));
-    args.push_back(std::to_string(ishape[1]));
-    args.push_back(std::to_string(wshape[0]));
+  GenerateBodyOutput GenerateOpCall(const CallNode* call) {
+    const auto* op_node = call->op.as<OpNode>();
+    CHECK(op_node) << "OpNode expected, got something else";
```
Comment (zhiics) [inline src/relay/backend/contrib/dnnl/codegen.cc:231]: Let's also print out what the op_node is
```
CHECK(op_node) << "Expect OpNode, but got " << op_node->GetTypeKey();
```

File: tests/python/relay/test_pass_partition_graph.py (hunk 1 above)
Comment (zhiics) [reply tests/python/relay/test_pass_partition_graph.py:901]: Yes, some common patterns could be moved to dnnl.py. Let's keep it here for now and have a separate PR to do that.

File: src/relay/backend/contrib/dnnl/codegen.cc (hunk 3 above)
Comment (masahi) [reply src/relay/backend/contrib/dnnl/codegen.cc:54]: yes exactly. I'm working on improving it right now, to support conv2d + relu pattern. I'm adding "depth" and "expected_op_names" arguments.

File: src/relay/backend/contrib/dnnl/codegen.cc (hunk 4 above)
Comment (masahi) [reply src/relay/backend/contrib/dnnl/codegen.cc:231]: You mean `call->GetTypeKey()` right? When this check fails, op_node is nullptr.

Comment (zhiics) [reply src/relay/backend/contrib/dnnl/codegen.cc:231]: yes, my bad.

File: src/relay/backend/contrib/dnnl/codegen.cc (hunk 3 above)
Comment (masahi) [reply src/relay/backend/contrib/dnnl/codegen.cc]: fixed, please see the new change

Comment (alexbooth) [reply src/relay/backend/contrib/dnnl/codegen.cc]: Sounds good to me. Backend I'm working on can use this, so it would be nice if this can move to utils.h where you moved IsOp and GetShape.

**Comment by masahi** [comment]
Also added support for conv2d + relu pattern. This shows how easy to add a new pattern and support it in the codegen.

File: src/relay/backend/contrib/dnnl/codegen.cc (hunk 3 above)
Comment (masahi) [reply src/relay/backend/contrib/dnnl/codegen.cc]: Ok done. See the last commit.

Comment (alexbooth) [reply src/relay/backend/contrib/dnnl/codegen.cc]: Thanks

File: src/relay/backend/utils.h [hunk 5]
```diff
@@ -152,8 +152,35 @@ inline bool IsOp(const CallNode* call, const std::string& op_name) {
...
+    CHECK(IsOp(current_call, "nn.conv2d"));
+    return current_call;
+  }
+
+  CHECK(depth < expected_op_names.size() && IsOp(current_call, expected_op_names[depth]));
+  CHECK_GT(current_call->args.size(), 0);
+
+  const auto* next_call = current_call->args[0].as<CallNode>();
+  return GetRootConv2DCall(next_call, depth - 1, expected_op_names);
+}
```
Comment (alexbooth) [inline src/relay/backend/utils.h:179]: I'm not incredibly familiar with majority of relay ops, but would it be safe to also modify this "GetRootCall" in the case where the root composite call is something other than "nn.conv2d" ("nn.dense" for example) ?

Comment (masahi) [reply src/relay/backend/utils.h:179]: ah thanks that is a good idea since nothing there is specific to conv2d. We could also make `expected_op_names` arg optional, so that this can be used to retrieve any ancestor node at height `depth` above from `current_call`. If this sounds useful outside of fused call use cases, I can make it more general like that. What do you think? I can certainly make it work with dense or other `root_op`. cc @zhiics

Comment (masahi) [reply src/relay/backend/utils.h]: Renamed to `"GetRootCall` and removed the check of `nn.conv2d` so that it can be used for other root ops. `expected_op_names` is required. Please have a look

Comment (alexbooth) [reply src/relay/backend/utils.h]: Great! Yes this will work for majority of my use cases. If there is an easier way to find a composite function's root call, maybe someone else can comment.

**Comment by masahi** [comment]
@tqchen Can you have a look at CI problem? https://ci.tvm.ai/blue/organizations/jenkins/tvm/detail/PR-5272/7/pipeline/152

**Comment by tqchen** [comment]
Please try to retrigger https://github.com/apache/incubator-tvm/issues/5285

**Comment by masahi** [comment]
Yes I got that error too, but also there is a problem on i386 CI causing segfault from crt. https://ci.tvm.ai/blue/organizations/jenkins/tvm/detail/PR-5272/7/pipeline/152. This PR https://github.com/apache/incubator-tvm/pull/5280 also had the same segfault from i386.

File: tests/python/relay/test_pass_partition_graph.py [hunk 6]
```diff
@@ -856,6 +857,128 @@ def expected():
...
+
+def test_dnnl_fuse():
+    def make_pattern(with_bias=True):
+        data = relay.var("data", relay.TensorType((1, 3, 224, 224), "float32"))
+        weight = relay.var("weight")
+        bias = relay.var("bias")
+        conv = relay.nn.conv2d(data=data, weight=weight, kernel_size=(3, 3),
+                               channels=8, padding=(1, 1))
+        if with_bias:
+            conv_out = relay.add(conv, bias)
```
Comment (alexbooth) [inline tests/python/relay/test_pass_partition_graph.py:869]: Should this be ```relay.nn.bias_add```? On my side, I had ```nn.bias_add``` after conv2d and noticed that the GetRootCall would fail unless I changed ```add``` to ```nn.bias_add```.

File: tests/python/relay/test_pass_partition_graph.py (hunk 2 above)
Comment (alexwong) [reply tests/python/relay/test_pass_partition_graph.py:918]: Any updates / initial findings?

Comment (comaniac) [reply tests/python/relay/test_pass_partition_graph.py:918]: Hey I just tried your case with #5277 and it works fine. Maybe we can push to merge that PR so that this problem will be gone naturally.

Comment (zhiics) [reply tests/python/relay/test_pass_partition_graph.py:918]: Looks like that is caused by wrongly adding the `default` compiler annotations. Let's work on reviewing #5277 and merge these bunch of fixes.

**Comment by masahi** [comment]
> Although I still feel this is a bit ad-hoc, it should be fine for now as we only use DNNL as an example to demonstrate external codegen.

What is still ad-hoc? I think I explained the motivation clearly in the original PR #4741 and the current impl using composite is the best we could do.

File: tests/python/relay/test_pass_partition_graph.py (hunk 6 above)
Comment (masahi) [reply tests/python/relay/test_pass_partition_graph.py:869]: No, since I use this pattern to detect conv + add + relu that come from decomposing the batch norm.  `SimplyfyInference` generates `relay.add`, see https://github.com/apache/incubator-tvm/blob/14ba49c60c49474a564f990363de9d114c9b019b/src/relay/transforms/simplify_inference.cc#L54

We can have both conv2d + add + relu and conv2d + bias_add + relu patterns in the table.

Comment (alexbooth) [reply tests/python/relay/test_pass_partition_graph.py:869]: Understood, thank you.

**Comment by comaniac** [comment]
> What is still ad-hoc? I think I explained the motivation clearly in the original PR #4741 and the current impl using composite is the best we could do. Wanting fused op support in external codegen is a natural and @alexbooth already finds it useful.

Sorry for being unclear. The motivation of using fused ops to demonstrate composite functions is definitely clear. What I meant was the general implementation of DNNL codegen as I've mentioned in #4741, so I'm good with this PR.

File: tests/python/relay/test_pass_partition_graph.py (hunk 2 above)
Comment (alexwong) [reply tests/python/relay/test_pass_partition_graph.py:920]: Yes, this PR fixed it!

**Comment by zhiics** [comment]
@masahi please rebase when you get a chance

**Comment by zhiics** [comment]
@masahi just a reminder, probably don't rebase again the master, other changes are here as well

**Comment by masahi** [comment]
Yes I'm working on fixing it. Resolving conflict turned out a bit complicated, so I wanted to start fresh.

File: src/relay/backend/contrib/dnnl/codegen.cc [hunk 7]
```diff
@@ -53,163 +141,186 @@ class CodegenDNNL : public ExprVisitor, public CodegenCBase {
   }
 
   void VisitExpr_(const TupleGetItemNode* op) final {
-    // Do nothing
+    VisitExpr(op->tuple);
```
Comment (masahi) [inline src/relay/backend/contrib/dnnl/codegen.cc:144]: I wondered why this change is on my diff, but it seems TupleGetItemNode handling was reverted in #5277

**Comment by masahi** [comment]
@zhiics @comaniac The change to the vm/compiler.cc that moved second fusion before Inline has been reverted in #5277. Should I bring it back in this PR?

**Comment by comaniac** [comment]
> @zhiics @comaniac The change to the vm/compiler.cc that moved second fusion before Inline has been reverted in #5277. Should I bring it back in this PR?

Hey sorry for that. Yes that was a mistake. When I was rebasing to the master, the commit that decouples DNNL and VM part was reapplied, so #5288 was reverted as you pointed out. If possible, would you cherry-pick it back in this PR? Thanks!

**Comment by masahi** [comment]
@comaniac Done. Please take a look at the last commit

**Comment by masahi** [comment]
Changes to multi output support in dnnl/codegen.cc has also been reverted in #5277, but I've already integrated that change in this PR last night.

**Comment by comaniac** [comment]
> Changes to multi output support in dnnl/codegen.cc has also been reverted in #5277, but I've already integrated that change in this PR last night.

Yes I've also checked that those diffs are in this PR as well. Thanks.

**Comment by masahi** [comment]
Thanks @zhiics @comaniac
